fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OutputPropertyDeriver.java

```diff
@@ -0,0 +1,550 @@
+import com.starrocks.sql.optimizer.operator.physical.PhysicalFilterOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalHashAggregateOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalHashJoinOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalLimitOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalNoCTEOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalOlapScanOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalTableFunctionOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalTopNOperator;
+import com.starrocks.sql.optimizer.operator.physical.PhysicalWindowOperator;
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.rule.transformation.JoinPredicateUtils;
+import com.starrocks.sql.optimizer.task.TaskContext;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.starrocks.sql.optimizer.rule.transformation.JoinPredicateUtils.getEqConj;
+
+public class OutputPropertyDeriver extends OperatorVisitor<PhysicalPropertySet, ExpressionContext> {
+    private GroupExpression groupExpression;
+    private PhysicalPropertySet requirements;
+    // children best group expression
+    private List<GroupExpression> childrenBestExprList;
+    // input(required) property for children.
+    private List<PhysicalPropertySet> childrenInputProperties;
+    // children output property
+    private List<PhysicalPropertySet> childrenOutputProperties;
+    private double curTotalCost;
+    private final OptimizerContext context;
+
+    public OutputPropertyDeriver(TaskContext taskContext) {
+        this.context = taskContext.getOptimizerContext();
+    }
+
+    public PhysicalPropertySet getOutputProperty(
+            PhysicalPropertySet requirements,
+            GroupExpression groupExpression,
+            List<GroupExpression> childrenBestExprList,
+            List<PhysicalPropertySet> childrenInputProperties,
+            List<PhysicalPropertySet> childrenOutputProperties) {
+        this.requirements = requirements;
+        this.groupExpression = groupExpression;
+        this.childrenBestExprList = childrenBestExprList;
+        this.childrenInputProperties = childrenInputProperties;
+        this.childrenOutputProperties = childrenOutputProperties;
+
+        return groupExpression.getOp().accept(this, new ExpressionContext(groupExpression));
+    }
+
+    public Pair<PhysicalPropertySet, Double> getOutputPropertyWithCost(
+            PhysicalPropertySet requirements,
+            GroupExpression groupExpression,
+            List<GroupExpression> childrenBestExprList,
+            List<PhysicalPropertySet> childrenInputProperties,
```

Comment:
```suggestion
            List<PhysicalPropertySet> inputProperties,
```
It's not InputProperties of children